File: app/src/main/java/me/thanish/prayers/se/domain/PrayerTimeTable.kt
package me.thanish.prayers.se.domain

import android.content.Context
import java.time.LocalDate

/**
 * PrayerTimeTable is a data class representing a prayer time table.
 * This contains all the prayer times for a given date and given city.
 */
data class PrayerTimeTable(
    val city: PrayerTimeCity,
    val date: LocalDate,
    val fajr: PrayerTime,
    val shuruk: PrayerTime,
    val dhohr: PrayerTime,
    val asr: PrayerTime,
    val maghrib: PrayerTime,
    val isha: PrayerTime,
) {
    /**
     * toList returns a list of prayer times in order.
     */
    fun toList(): List<PrayerTime> {
        return listOf(fajr, shuruk, dhohr, asr, maghrib, isha)
    }

    /**
     * ✄ - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
     */

    companion object {
        /**
         * fromAssetData creates a prayer time table from asset data.
         */
        fun forDate(context: Context, city: PrayerTimeCity, date: LocalDate): PrayerTimeTable {
            val method = PrayerTimeMethod.get(context)
            val data = PrayerTimeData.get(context, city)

            return PrayerTimeTable(
                city,
                date,
                PrayerTime(city, PrayerTimeType.fajr, data.getFajr(date)),
                PrayerTime(city, PrayerTimeType.shuruk, data.getShuruk(date)),
                PrayerTime(city, PrayerTimeType.dhohr, data.getDhohr(date)),
                PrayerTime(city, PrayerTimeType.asr, data.getAsr(date, method)),
                PrayerTime(city, PrayerTimeType.maghrib, data.getMaghrib(date)),
                PrayerTime(city, PrayerTimeType.isha, data.getIsha(date)),
            )
        }

        /**
         * forToday creates a prayer time table for today.
         */
        fun forToday(context: Context, city: PrayerTimeCity): PrayerTimeTable {
            return forDate(context, city, LocalDate.now())
        }
    }
}

File: apps/iphone/Bönetider/Bönetider/Domain/PrayerTimeTable.swift
//
//  PrayerTimeTable.swift
//  Bönetider
//
//  Created by Thanish Nizam on 2025-01-31.
//


import Foundation

// PrayerTimeTable is a helper struct that makes it easy to bundle prayer times
// for a specific date together. Provides computed properties for easy access.
struct PrayerTimeTable {
  // Get the PrayerTimeTable for given date and city.
  static func forDate(city: PrayerTimeCity, date: Date) -> PrayerTimeTable {
    let values = PrayerTime.getPrayersForDate(city: city, date: date)
    return PrayerTimeTable.init(city: city, date: date, data: values)
  }
  
  // MARK: - Properties
  
  let city: PrayerTimeCity
  let date: Date
  let data: [PrayerTime]
  
  // MARK: - Computed Properties
  
  var fajr: PrayerTime { data.first(where: { $0.type == .fajr })! }
  var shuruk: PrayerTime { data.first(where: { $0.type == .shuruk })! }
  var dhohr: PrayerTime { data.first(where: { $0.type == .dhohr })! }
  var asr: PrayerTime { data.first(where: { $0.type == .asr })! }
  var maghrib: PrayerTime { data.first(where: { $0.type == .maghrib })! }
  var isha: PrayerTime { data.first(where: { $0.type == .isha })! }
  
  // MARK: - Initializers
  
  init(city: PrayerTimeCity, date: Date, data: [PrayerTime]) {
    self.city = city
    self.date = date
    self.data = data
  }
}
